mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp

```diff
@@ -679,6 +863,276 @@ mlir::scf::tileUsingSCF(RewriterBase &rewriter, TilingInterface op,
+    // the original op because it could lead to
+    // `rewriter.replaceOp(op, op->getResults())` and user would get crash.
+    if (llvm::all_of(tileSizeVector, isZeroIndex)) {
+      return scf::SCFTilingResult{/*tiledOps=*/{clonedOp}, /*loops=*/{},
+                                  clonedOp->getResults()};
+    }
+
+    // 5c. Tile the cloned operation.
+    FailureOr<TilingResult> tiledImplementation =
+        clonedOp.getTiledImplementation(rewriter, offsets, sizes);
+    if (failed(tiledImplementation)) {
+      return rewriter.notifyMatchFailure(op, "failed to tile operation");
+    }
+
+    origTiledImplementation.tiledOps.append(tiledImplementation->tiledOps);
+
+    // 5d. Delete the cloned operation.
+    rewriter.eraseOp(clonedOp);
+
+    // If loops are empty, the tiled op is used as the replacement for the
+    // untiled op.
+    if (forLoops.empty()) {
+      return scf::SCFTilingResult{tiledImplementation->tiledOps, forLoops,
+                                  tiledImplementation->tiledValues};
+    }
+
+    if (op->getNumResults() == 0) {
+      // The innermost loop does not have a `scf.yield` yet. There is nothing to
+      // return, so generate an empty `scf.yield` operation.
+      rewriter.setInsertionPointToEnd(innermostLoop.getBody());
+      rewriter.create<scf::YieldOp>(op->getLoc());
+    }
+
+    // 6. Yield all the results of the tiled operation.
+    int64_t numResults = op->getNumResults();
+    SmallVector<SmallVector<OpFoldResult>> resultOffsetsList(numResults),
+        resultSizesList(numResults);
+    SmallVector<Value> yieldedValues;
+    for (auto [index, tiledValue] :
+         llvm::enumerate(tiledImplementation->tiledValues)) {
+      SmallVector<OpFoldResult> resultOffsets, resultSizes;
+      if (failed(op.getResultTilePosition(rewriter, index, offsets, sizes,
+                                          resultOffsets, resultSizes))) {
+        return rewriter.notifyMatchFailure(
+            op, "failed to get slice of result produced");
+      }
+      SmallVector<OpFoldResult> resultStrides(resultOffsets.size(),
+                                              rewriter.getIndexAttr(1));
+      auto insertSlice = rewriter.create<tensor::InsertSliceOp>(
+          op->getLoc(), tiledValue, clonedOpDestination[index], resultOffsets,
+          resultSizes, resultStrides);
+      yieldedValues.push_back(insertSlice);
+    }
+    rewriter.create<scf::YieldOp>(op->getLoc(), yieldedValues);
+  }
+
+  SmallVector<Value> replacements = llvm::map_to_vector(
+      forLoops.front()->getResults(), [](OpResult r) -> Value { return r; });
+  LLVM_DEBUG({
+    if (!forLoops.empty()) {
+      llvm::dbgs() << "After tiled implementation :\n";
+      forLoops.front().dump();
+      llvm::dbgs() << "\n";
+    }
+  });
+
+  /// Remove outermost tailend loop as its only use was to compute
+  /// replacements.
+  forLoops.erase(forLoops.begin());
+
```

Comment:
Not quite sure what happens here: isn't this loop produced by tiling and therefore should appear in the results?